lib/matplotlib/tight_layout.py

```diff
@@ -106,66 +77,39 @@ def auto_adjust_subplotpars(
         if all(not ax.get_visible() for ax in subplots):
             continue
 
-        tight_bbox_raw = union([ax.get_tightbbox(renderer) for ax in subplots
-                                if ax.get_visible()])
+        tight_bbox_raw = Bbox.union([
+            ax.get_tightbbox(renderer) for ax in subplots if ax.get_visible()])
         tight_bbox = TransformedBbox(tight_bbox_raw,
                                      fig.transFigure.inverted())
 
         row1, col1 = divmod(num1, cols)
-
         if num2 is None:
-            # left
-            hspaces[row1 * (cols + 1) + col1].append(
-                                        _get_left(tight_bbox, ax_bbox))
-            # right
-            hspaces[row1 * (cols + 1) + (col1 + 1)].append(
-                                        _get_right(tight_bbox, ax_bbox))
-            # top
-            vspaces[row1 * cols + col1].append(
-                                        _get_top(tight_bbox, ax_bbox))
-            # bottom
-            vspaces[(row1 + 1) * cols + col1].append(
-                                        _get_bottom(tight_bbox, ax_bbox))
+            num2 = num1
+        row2, col2 = divmod(num2, cols)
 
-        else:
-            row2, col2 = divmod(num2, cols)
-
-            for row_i in range(row1, row2 + 1):
-                # left
-                hspaces[row_i * (cols + 1) + col1].append(
-                                    _get_left(tight_bbox, ax_bbox))
-                # right
-                hspaces[row_i * (cols + 1) + (col2 + 1)].append(
-                                    _get_right(tight_bbox, ax_bbox))
-            for col_i in range(col1, col2 + 1):
-                # top
-                vspaces[row1 * cols + col_i].append(
-                                    _get_top(tight_bbox, ax_bbox))
-                # bottom
-                vspaces[(row2 + 1) * cols + col_i].append(
-                                    _get_bottom(tight_bbox, ax_bbox))
+        for row_i in range(row1, row2 + 1):
+            hspaces[row_i, col1] += ax_bbox.xmin - tight_bbox.xmin  # left
+            hspaces[row_i, col2 + 1] += tight_bbox.xmax - ax_bbox.xmax  # right
+        for col_i in range(col1, col2 + 1):
+            vspaces[row1, col_i] += tight_bbox.ymax - ax_bbox.ymax  # top
+            vspaces[row2 + 1, col_i] += ax_bbox.ymin - tight_bbox.ymin  # bot.
 
     fig_width_inch, fig_height_inch = fig.get_size_inches()
 
-    # margins can be negative for axes with aspect applied. And we
-    # append + [0] to make minimum margins 0
-
+    # margins can be negative for axes with aspect applied, so use max(, 0) to
+    # make them nonnegative.
     if not margin_left:
-        margin_left = max([sum(s) for s in hspaces[::cols + 1]] + [0])
-        margin_left += pad_inches / fig_width_inch
-
+        margin_left = (max(hspaces[:, 0].max(), 0)
```

Comment:
What did the sum do? And why don‘t we need it anymore?

Reply:
For reasons I don't understand, this used to sum the margins of all subplots at a given position -- I'd have thought it's the max that matters, but let's not touch that here. (Actually I don't even know how you can realistically have multiple subplots at the same position, but there's a comment specifically mentioning this happens in axes_grid, so sure...)
Previously the code accumulated all margins in lists, and summed them at the end (`hspaces[...].append(...)`).  Now I just make a 2D array of floats and directly sum the margins there (`hspaces[...] += ...`).

Reply:
Actually, looking at it again, I see at least one reason for the sum: you need to add the right margin (between the spines and the label edges) of the axes on the left and the left margin of the axes on the right.